File: agents/linkedin/oauth-server.ts
/**
 * LinkedIn OAuth 2.0 Server
 *
 * Run once to authenticate and save tokens:
 *   bun run oauth-server.ts
 *
 * Opens browser, catches callback, saves tokens to .env
 */

import { $ } from "bun";

const CLIENT_ID = process.env.LINKEDIN_CLIENT_ID;
const CLIENT_SECRET = process.env.LINKEDIN_CLIENT_SECRET;
const REDIRECT_URI = "http://localhost:3000/callback";
const SCOPES = ["openid", "profile", "w_member_social"].join(" ");

if (!CLIENT_ID || !CLIENT_SECRET) {
  console.error("Missing LINKEDIN_CLIENT_ID or LINKEDIN_CLIENT_SECRET in environment");
  console.error("Create a .env file with these values first");
  process.exit(1);
}

const authUrl = new URL("https://www.linkedin.com/oauth/v2/authorization");
authUrl.searchParams.set("response_type", "code");
authUrl.searchParams.set("client_id", CLIENT_ID);
authUrl.searchParams.set("redirect_uri", REDIRECT_URI);
authUrl.searchParams.set("scope", SCOPES);
authUrl.searchParams.set("state", crypto.randomUUID());

console.log("\n🔐 LinkedIn OAuth Flow\n");
console.log("Opening browser for authorization...\n");

// Open browser
await $`xdg-open ${authUrl.toString()}`.quiet();

// Start server to catch callback
const server = Bun.serve({
  port: 3000,
  async fetch(req) {
    const url = new URL(req.url);

    if (url.pathname === "/callback") {
      const code = url.searchParams.get("code");
      const error = url.searchParams.get("error");

      if (error) {
        console.error("❌ Authorization denied:", error);
        server.stop();
        return new Response(`Authorization denied: ${error}`, { status: 400 });
      }

      if (!code) {
        return new Response("No code received", { status: 400 });
      }

      console.log("✅ Authorization code received, exchanging for tokens...\n");

      // Exchange code for tokens
      const tokenResponse = await fetch("https://www.linkedin.com/oauth/v2/accessToken", {
        method: "POST",
        headers: {
          "Content-Type": "application/x-www-form-urlencoded",
        },
        body: new URLSearchParams({
          grant_type: "authorization_code",
          code,
          client_id: CLIENT_ID!,
          client_secret: CLIENT_SECRET!,
          redirect_uri: REDIRECT_URI,
        }),
      });

      if (!tokenResponse.ok) {
        const errorText = await tokenResponse.text();
        console.error("❌ Token exchange failed:", errorText);
        server.stop();
        return new Response(`Token exchange failed: ${errorText}`, { status: 500 });
      }

      const tokens = await tokenResponse.json();

      console.log("✅ Tokens received!\n");
      console.log("Access Token:", tokens.access_token?.substring(0, 20) + "...");
      console.log("Expires in:", tokens.expires_in, "seconds");

      // Check for id_token (JWT with user info)
      let userId = "";
      if (tokens.id_token) {
        console.log("ID Token: present");
        // Decode JWT payload (middle part)
        const payload = JSON.parse(Buffer.from(tokens.id_token.split('.')[1], 'base64').toString());
        console.log("User Sub:", payload.sub);
        console.log("Name:", payload.name);
        userId = payload.sub;
      }

      // Save to .env
      const envContent = `# LinkedIn OAuth Tokens - Generated ${new Date().toISOString()}
LINKEDIN_CLIENT_ID=${CLIENT_ID}
LINKEDIN_CLIENT_SECRET=${CLIENT_SECRET}
LINKEDIN_ACCESS_TOKEN=${tokens.access_token}
LINKEDIN_REFRESH_TOKEN=${tokens.refresh_token || ""}
LINKEDIN_TOKEN_EXPIRES=${Date.now() + (tokens.expires_in * 1000)}
LINKEDIN_USER_SUB=${userId}
`;

      await Bun.write(".env", envContent);
      console.log("\n✅ Tokens saved to .env\n");
      console.log("You can now run: bun run test-connection.ts");

      server.stop();
      return new Response(`
        <html>
          <body style="font-family: system-ui; padding: 40px; text-align: center;">
            <h1>✅ Authorization Successful!</h1>
            <p>Tokens have been saved. You can close this window.</p>
          </body>
        </html>
      `, { headers: { "Content-Type": "text/html" } });
    }

    return new Response("LinkedIn OAuth Server", { status: 200 });
  },
});

console.log(`Server running at http://localhost:${server.port}`);
console.log("Waiting for callback...\n");

File: agents/linkedin/oauth_server.py
"""LinkedIn OAuth 2.0 server for Content Engine."""

import os
import sys
import webbrowser
from http.server import HTTPServer, BaseHTTPRequestHandler
from pathlib import Path
from urllib.parse import urlparse, parse_qs
import requests

from lib.config import get_linkedin_config, get_server_config
from lib.errors import OAuthError
from lib.logger import setup_logger


logger = setup_logger(__name__)


class OAuthHandler(BaseHTTPRequestHandler):
    """HTTP handler for OAuth callback."""

    auth_code: str | None = None

    def log_message(self, format: str, *args) -> None:
        """Override to use custom logger."""
        logger.info(format % args)

    def do_GET(self) -> None:
        """Handle OAuth callback GET request."""
        parsed_url = urlparse(self.path)

        if parsed_url.path == "/":
            # Initial page - redirect to LinkedIn
            config = get_linkedin_config()
            auth_url = (
                "https://www.linkedin.com/oauth/v2/authorization"
                f"?response_type=code"
                f"&client_id={config.client_id}"
                f"&redirect_uri={config.redirect_uri}"
                f"&scope=openid%20profile%20w_member_social"
            )

            self.send_response(302)
            self.send_header("Location", auth_url)
            self.end_headers()

        elif parsed_url.path == "/auth/callback":
            # OAuth callback with authorization code
            query_params = parse_qs(parsed_url.query)

            if "error" in query_params:
                error = query_params["error"][0]
                logger.error(f"OAuth error: {error}")
                self.send_response(400)
                self.send_header("Content-type", "text/html")
                self.end_headers()
                self.wfile.write(f"<h1>Error: {error}</h1>".encode())
                return

            if "code" not in query_params:
                logger.error("No authorization code in callback")
                self.send_response(400)
                self.send_header("Content-type", "text/html")
                self.end_headers()
                self.wfile.write(b"<h1>Error: No authorization code</h1>")
                return

            OAuthHandler.auth_code = query_params["code"][0]
            logger.info(f"✅ Received authorization code: {OAuthHandler.auth_code[:20]}...")

            self.send_response(200)
            self.send_header("Content-type", "text/html")
            self.end_headers()
            self.wfile.write(
                b"<h1>Success!</h1>"
                b"<p>Authorization code received. You can close this window.</p>"
                b"<p>Check your terminal for next steps.</p>"
            )

        else:
            self.send_response(404)
            self.end_headers()


def update_env_file(access_token: str, user_sub: str) -> None:
    """
    Update .env file with new access token and user sub.

    Args:
        access_token: LinkedIn access token
        user_sub: LinkedIn user sub identifier
    """
    env_path = Path(".env")

    if not env_path.exists():
        logger.warning(".env file not found, creating new one")
        env_path.touch()

    # Read existing .env
    lines = env_path.read_text().splitlines()

    # Update or add tokens
    updated_lines = []
    token_found = False
    sub_found = False

    for line in lines:
        if line.startswith("LINKEDIN_ACCESS_TOKEN="):
            updated_lines.append(f"LINKEDIN_ACCESS_TOKEN={access_token}")
            token_found = True
        elif line.startswith("LINKEDIN_USER_SUB="):
            updated_lines.append(f"LINKEDIN_USER_SUB={user_sub}")
            sub_found = True
        else:
            updated_lines.append(line)

    # Add if not found
    if not token_found:
        updated_lines.append(f"LINKEDIN_ACCESS_TOKEN={access_token}")
    if not sub_found:
        updated_lines.append(f"LINKEDIN_USER_SUB={user_sub}")

    # Write back
    env_path.write_text("\n".join(updated_lines) + "\n")
    logger.info("✅ Updated .env file with new credentials")


def exchange_code_for_token(auth_code: str) -> tuple[str, str]:
    """
    Exchange authorization code for access token.

    Args:
        auth_code: OAuth authorization code

    Returns:
        Tuple of (access_token, user_sub)

    Raises:
        OAuthError: If token exchange fails
    """
    config = get_linkedin_config()

    logger.info("Exchanging authorization code for access token...")

    try:
        response = requests.post(
            "https://www.linkedin.com/oauth/v2/accessToken",
            data={
                "grant_type": "authorization_code",
                "code": auth_code,
                "client_id": config.client_id,
                "client_secret": config.client_secret,
                "redirect_uri": config.redirect_uri,
            },
            timeout=30,
        )

        if not response.ok:
            raise OAuthError(f"Token exchange failed: {response.status_code} {response.text}")

        token_data = response.json()
        access_token = token_data["access_token"]

        logger.info("✅ Access token received")

        # Get user info (for user_sub)
        user_response = requests.get(
            "https://api.linkedin.com/v2/userinfo",
            headers={"Authorization": f"Bearer {access_token}"},
            timeout=30,
        )

        if not user_response.ok:
            raise OAuthError(f"User info request failed: {user_response.status_code}")

        user_data = user_response.json()
        user_sub = user_data["sub"]

        logger.info(f"✅ User sub received: {user_sub}")

        return access_token, user_sub

    except requests.RequestException as e:
        raise OAuthError(f"Request failed: {str(e)}")


def main() -> None:
    """Run OAuth server."""
    logger.info("=" * 60)
    logger.info("LinkedIn OAuth 2.0 Server")
    logger.info("=" * 60)

    server_config = get_server_config()
    linkedin_config = get_linkedin_config()

    server_address = (server_config.host, server_config.port)
    httpd = HTTPServer(server_address, OAuthHandler)

    logger.info(f"\n🌐 Server running at http://localhost:{server_config.port}")
    logger.info("Opening browser for LinkedIn authorization...")
    logger.info("Waiting for OAuth callback...\n")

    # Open browser
    webbrowser.open(f"http://localhost:{server_config.port}")

    # Handle requests until we get the auth code
    while OAuthHandler.auth_code is None:
        httpd.handle_request()

    httpd.server_close()

    # Exchange code for token
    try:
        access_token, user_sub = exchange_code_for_token(OAuthHandler.auth_code)

        logger.info("\n" + "=" * 60)
        logger.info("🎉 OAuth flow completed successfully!")
        logger.info("=" * 60)

        # Auto-save to .env
        try:
            update_env_file(access_token, user_sub)
            logger.info("\n✅ Credentials saved to .env file automatically!")
            logger.info("\nYou can now use:")
            logger.info("  uv run python -m agents.linkedin.test_connection")
            logger.info("  uv run content-engine draft 'Your post'\n")
        except Exception as e:
            logger.warning(f"⚠️  Failed to auto-save to .env: {e}")
            logger.info("\nManually add these to your .env file:")
            logger.info(f"\nLINKEDIN_ACCESS_TOKEN={access_token}")
            logger.info(f"LINKEDIN_USER_SUB={user_sub}\n")

    except OAuthError as e:
        logger.error(f"❌ OAuth error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()
